pkg/stream/stream.go

```diff
@@ -0,0 +1,336 @@
+//go:generate mockgen  -destination=./mocks/stream.go -package=mocks github.com/blocknative/dreamboat/pkg/stream Pubsub,RemoteDatastore
+package stream
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/flashbots/go-boost-utils/types"
+	"github.com/lthibault/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/blocknative/dreamboat/pkg/datastore"
+	"github.com/blocknative/dreamboat/pkg/structs"
+)
+
+type Pubsub interface {
+	Publish(context.Context, string, []byte) error
+	Subscribe(context.Context, string) (chan []byte, error)
+}
+
+type RemoteDatastore interface {
+	GetPayload(context.Context, structs.PayloadKey) (*structs.BlockAndTrace, error)
+	PutPayload(context.Context, structs.PayloadKey, *structs.BlockAndTrace, time.Duration) error
+}
+
+type StreamConfig struct {
+	Logger          log.Logger
+	ID              string
+	TTL             time.Duration
+	PubsubTopic     string // pubsub topic name for block submissions
+	PublishAll      bool
+	StreamQueueSize int
+}
+
+type StreamDatastore struct {
+	*datastore.Datastore
+	Pubsub          Pubsub
+	RemoteDatastore RemoteDatastore
+
+	cacheRequests chan *structs.BlockAndTrace
+	storeRequests chan *structs.BlockAndTrace
+
+	Config StreamConfig
+	Logger log.Logger
+
+	m StreamMetrics
+}
+
+func NewStreamDatastore(ds *datastore.Datastore, ps Pubsub, rds RemoteDatastore, cfg StreamConfig) *StreamDatastore {
+	s := StreamDatastore{
+		Datastore:       ds,
+		Pubsub:          ps,
+		RemoteDatastore: rds,
+		cacheRequests:   make(chan *structs.BlockAndTrace, cfg.StreamQueueSize),
+		storeRequests:   make(chan *structs.BlockAndTrace, cfg.StreamQueueSize),
+		Config:          cfg,
+		Logger:          cfg.Logger.WithField("relay-service", "stream"),
+	}
+
+	s.initMetrics()
+
+	return &s
+}
+
+func (s *StreamDatastore) RunSubscriber(ctx context.Context) error {
+	blocks, err := s.Pubsub.Subscribe(ctx, s.Config.PubsubTopic)
+	if err != nil {
+		return err
+	}
+
+	sBlock := StreamBlock{}
+	for rawSBlock := range blocks {
+		if err := proto.Unmarshal(rawSBlock, &sBlock); err != nil {
+			s.Logger.Warnf("fail to decode stream block: %s", err.Error())
```

Comment:
Same as above:  I recommend using `WithError` and avoiding string-formatting in the message.  There's a bunch of these in this PR, so I won't flag every single one :)